django/db/models/indexes.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import unicode_literals
+
+import hashlib
+
+from django.utils.encoding import force_bytes
+
+__all__ = ['Index']
+
+# The max length of the names of the indexes (restricted to 30 due to Oracle)
+MAX_NAME_LENGTH = 30
+
+
+class Index(object):
+
+    index_type = 'idx'
+
+    def __init__(self, fields=[], name=None):
+        if not fields:
+            raise ValueError('At least one field is required to define an index.')
+        self.fields = fields
+        self._name = name or ''
+        if self._name:
+            self.normalise_name()
+            if len(self._name) > MAX_NAME_LENGTH:
+                raise ValueError("Index names cannot be longer than %s characters." % MAX_NAME_LENGTH)
+
+    @property
+    def name(self):
+        if not self._name:
+            self._name = self.get_name()
+            self.normalise_name()
+        return self._name
+
+    def normalise_name(self):
```

Comment:
I was expecting to raise a ValueError if these conditions aren't met, similar to what we do with "Index names cannot be longer". I don't think it's a good idea to modify the user provided value as it seems like that would only cause confusion.